pkg/controllers/vdb/crashloop_reconciler.go

```diff
@@ -0,0 +1,119 @@
+}
+
+// MakeCrashLoopReconciler will build a reconcile actor for CrashLoopReconciler
+func MakeCrashLoopReconciler(vdbrecon *VerticaDBReconciler, log logr.Logger,
+	vdb *vapi.VerticaDB) controllers.ReconcileActor {
+	return &CrashLoopReconciler{
+		VRec: vdbrecon,
+		Log:  log.WithName("CrashLoopReconciler"),
+		VDB:  vdb,
+	}
+}
+
+// Reconcile will check for a crash loop in vclusterOps deployments.
+func (c *CrashLoopReconciler) Reconcile(ctx context.Context, _ *ctrl.Request) (ctrl.Result, error) {
+	// This reconciler is intended to find cases where the image does not have
+	// the nma, causing the pod to get into a CrashLoop backoff. So, if not
+	// deployed with vclusterOps we can skip this entirely.
+	if !vmeta.UseVClusterOps(c.VDB.Annotations) {
+		return ctrl.Result{}, nil
+	}
+
+	// We have to be careful about logging events about the wrong deployment type
+	// for other kinds of crash loops. If the deployment is wrong it should
+	// happen before the version annotation has been set, or for the case
+	// of upgrade its still set to the old version. Exit out if the version in
+	// the annotation, if present, supports vclusterOps.
+	vinf, ok := c.VDB.MakeVersionInfo()
+	if ok && vinf.IsEqualOrNewer(vapi.VcluseropsAsDefaultDeploymentMethodMinVersion) {
+		return ctrl.Result{}, nil
+	}
+
+	c.reconcileStatefulSets(ctx)
+	return ctrl.Result{}, nil
+}
+
+func (c *CrashLoopReconciler) reconcileStatefulSets(ctx context.Context) {
+	finder := iter.MakeSubclusterFinder(c.VRec.Client, c.VDB)
+	stss, err := finder.FindStatefulSets(ctx, iter.FindExisting|iter.FindSorted)
+	if err != nil {
+		// This reconciler is a best effort. It only tries to surface meaningful
+		// error messages based on the events it see. For this reason, no errors
+		// are emitted. We will log them then carry on to the next reconciler.
+		c.Log.Info("Failure detecting in CrashLoopReconciler. Will continue on", "err", err)
+		return
+	}
+	for i := range stss.Items {
+		sts := &stss.Items[i]
+		for j := int32(0); j < *sts.Spec.Replicas; j++ {
+			pn := names.GenPodNameFromSts(c.VDB, sts, j)
+			pod := &corev1.Pod{}
+			err = c.VRec.Client.Get(ctx, pn, pod)
+			if err != nil {
+				// Any error found during fetch are ignored. We will just go onto the next pod.
+				continue
+			}
+			nmaStatus := c.findNMAContainerStatus(pod)
+			if nmaStatus == nil {
+				continue
+			}
+			if nmaStatus.RestartCount > 0 &&
+				!nmaStatus.Ready &&
+				nmaStatus.LastTerminationState.Terminated != nil &&
+				nmaStatus.LastTerminationState.Terminated.Reason == "StartError" {
+				c.VRec.Eventf(c.VDB, corev1.EventTypeWarning, events.WrongImage,
+					"Image cannot be used for vclusterOps deployments. Change the deployment by changing the %s annotation",
+					vmeta.VClusterOpsAnnotation)
+				// Don't bother checking anymore pods as this setting is global for the CR.
+				return
+			}
+		}
+	}
+}
+
+func (c *CrashLoopReconciler) findNMAContainerStatus(pod *corev1.Pod) *corev1.ContainerStatus {
```

Comment:
I see around some functions related to k8s, so I wonder if at some point a separate package will be needed, especially to avoid code duplication.

Reply:
Yes, we will have to keep an eye on it. I already moved this function out to the bulder package (even before I saw this) to use the same logic somewhere else.